File: kotlin/simple-api/src/main/kotlin/mtakeshi1/githut/io/model/MockDataAccess.kt
package mtakeshi1.githut.io.model

object MockDataAccess : DataAccess {

    private val products = listOf(Models.Product(1, "a", "aa", 10.0), Models.Product(2, "b", "bb", 20.0))

    private val orders = mutableListOf<Models.ShoppingCart>()

    override fun listProducts(): List<Models.Product> = products

    override fun productWithId(id: Int): Models.Product = products.filter { it.id == id }[0]

    override fun newOrder(items: List<Models.ShoppingCartEntry>): Models.ShoppingCart {
        val order =
            Models.ShoppingCart(orders.size, items.map { it.amount * productWithId(it.productId).price }.sum(), items)
        orders.add(order)
        return order
    }

    override fun orderWithId(id: Int): Models.ShoppingCart = orders.filter { it.id == id }[0]

}

File: scala/scala-rest/app/src/main/scala/simplerest/MockDataAccess.scala
package simplerest

import cats.effect.IO
import cats.implicits._
import scala.collection.mutable
import scala.collection.mutable.ListBuffer

object MockDataAccess extends DataAccess {

  private val products = List(Models.Product(0, "name", "desc", 1), Models.Product(1, "name1", "desc1", 2))
  private val orders = ListBuffer[Models.ShoppingCart]()

  override def listProducts(): IO[List[Models.ProductInformation]] = IO.pure(products.map(p => Models.ProductInformation(p.id, p.name)))

  override def productWithId(id: Int): IO[Models.Product] = IO.pure(products.find(_.id == id).get)

  override def newOrder(items: List[Models.ShoppingCartEntry]): IO[Models.ShoppingCart] = {
    for {
      prods <- items.map(i => productWithId(i.productId).map(prod => (prod, i.amount))).sequence
      total = prods.map(t => t._2 * t._1.price).sum
      order = Models.ShoppingCart(orders.size, total, items)
      _ = orders.addOne(order)
    } yield order
  }

  override def orderWithId(id: Int): IO[Models.ShoppingCart] = IO.pure(orders.find(_.id == id).get)
}
